OSDOCS#6673: Add initial OLM v1 subdir (#63086)
openshift/openshift-docs: modules/olmv1-catalog-plain.adoc

Reviewer: Four `.Procedure`s should = four modules right?

* link:https://github.com/opencontainers/image-spec#oci-image-format-specification[OCI image specification]
* link:https://github.com/operator-framework/rukpak/blob/main/docs/bundles/plain.md#image-source[RukPak > Building a plain bundle > Image source]
* link:https://github.com/operator-framework/rukpak/blob/main/docs/sources/image.md#private-image-registries[RukPak > Sources > Images > Private image registries]

[id="olmv1-creating-fbc_{context}"]
= Creating a file-based catalog

If you do not have a file-based catalog, you must perform the following steps to initialize the catalog.

.Procedure

. Create a directory for the catalog by running the following command:
+
[source,terminal]
----
$ mkdir <catalog_dir>
----

. In the same directory level, create a Dockerfile that can build a catalog image:
+
[NOTE]
====
The Dockerfile must be in the same parent directory as the catalog directory that you created in the previous step:

.Example directory structure
[source,terminal]
----
.
├── <catalog_dir>
└── <catalog_dir>.Dockerfile
----
====
+

> .Example `<catalog_dir>.Dockerfile`

Reviewer 2: This Dockerfile will not work since no `opm serve` in it. You can use the `opm generate dockerfile` to create one, such as
```console
[cloud-user@preserve-olm-env2 jian]$ opm generate dockerfile doc-catalog
[cloud-user@preserve-olm-env2 jian]$ cat doc-catalog.Dockerfile 
# The base image is expected to contain
# /bin/opm (with a serve subcommand) and /bin/grpc_health_probe
FROM quay.io/operator-framework/opm:latest

# Configure the entrypoint and command
ENTRYPOINT ["/bin/opm"]
CMD ["serve", "/configs", "--cache-dir=/tmp/cache"]

# Copy declarative config root into image at /configs and pre-populate serve cache
ADD doc-catalog /configs
RUN ["/bin/opm", "serve", "/configs", "--cache-dir=/tmp/cache", "--cache-only"]

# Set DC-specific label for the location of the DC root directory
# in the image
LABEL operators.operatorframework.io.index.configs.v1=/configs
```

+
[source,terminal]
----
$ opm init <extension_name> \
    --output json \
    > <catalog_dir>/index.json
----
+
This command generates an `olm.package` declarative config blob in the specified catalog configuration file.

[id="olmv1-adding-plain-bundle-to-fbc_{context}"]
= Adding a plain bundle to a file-based catalog

Currently, the `opm render` command does not support adding plain bundles to catalogs. You must manually add plain bundles to your file-based catalog, as shown in the following procedure.

.Procedure

. Verify that your catalog's `index.json` or `index.yaml` file is similar to the following example:
+
.Example `<catalog_dir>/index.json` file
[source,json]
----
{
    {
     "schema": "olm.package",
     "name": "<extension_name>",
    }
}
----

. To create an `olm.bundle` blob, edit your `index.json` or `index.yaml` file, similar to the following example:
+
.Example `<catalog_dir>/index.json` file with `olm.bundle` blob
[source,json]
----
{
   "schema": "olm.bundle",
    "name": "<extension_name>.v<version>",
    "package": "<extension_name>",
    "image": "quay.io/<organization_name>/<repository_name>:<image_tag>",

>     "properties": [

Reviewer 2: It's better to add the `olm.gvk` and `olm.bundle.object` types.

Reviewer 3: @jianzhangbjz why do we have to add those types? My assumption was that for plain bundle, they're not really required.

// Module included in the following assemblies:
//
// * operators/olm_v1/olmv1-managing-catalogs.adoc

ifdef::openshift-origin[]
:registry-image: quay.io/operator-framework/opm:latest
endif::[]
ifndef::openshift-origin[]
:registry-image: registry.redhat.io/openshift4/ose-operator-registry:v{product-version}
endif::[]

:_content-type: PROCEDURE

[id="olmv1-building-plain-bundle-image-source_{context}"]
= Building a plain bundle image from an image source

The Operator Controller currently supports installing plain bundles created only from a _plain bundle image_.

.Prequisites

* You have Kubernetes manifests for your bundle in a flat directory at the root of your project similar to the following structure:
+
.Example directory structure
[source,terminal]
----
manifests
├── namespace.yaml
├── service_account.yaml
├── cluster_role.yaml
├── cluster_role_binding.yaml
└── deployment.yaml
----

.Procedure

. At the root of your project, create a Dockerfile that can build a bundle image:
+
.Example `plainbundle.Dockerfile`

> [source,docker]

Reviewer: Fun new source tag for me :nerd_face:

----
FROM scratch <1>
ADD manifests /manifests
----
<1> Use the `FROM scratch` directive to make the size of the image smaller. No other files or directories are required in the bundle image.

> . Build an OCI-compliant image using your preferred build tool, similar to the following example:

Reviewer: Does OCI need a first use definition?

+
[source,terminal]
----
$ podman build -f plainbundle.Dockerfile -t \
    quay.io/<organization_name>/<repository_name>:<image_tag> . <1>
----
<1> Use an image tag that references a repository where you have push access privileges.

> . Push the image to your remote registry:

Reviewer: ```suggestion
. Push the image to your remote registry by running the following command:
```

.Procedure

. Create a directory for the catalog by running the following command:
+
[source,terminal]
----
$ mkdir <catalog_dir>
----

. Generate a Dockerfile that can build a catalog image by running the `opm generate dockerfile` command in the same directory level as the previous step:
+
[source,terminal,subs="attributes+"]
----
ifdef::openshift-origin[]
$ opm generate dockerfile <catalog_dir>
endif::[]
ifndef::openshift-origin[]
$ opm generate dockerfile <catalog_dir> \
    -i {registry-image} <1>
endif::[]
----
ifndef::openshift-origin[]
<1> Specify the official Red Hat base image by using the `-i` flag, otherwise the Dockerfile uses the default upstream image.
endif::[]
+
[NOTE]
====
The generated Dockerfile must be in the same parent directory as the catalog directory that you created in the previous step:

.Example directory structure
[source,terminal]
----
.
├── <catalog_dir>
└── <catalog_dir>.Dockerfile
----
====

. Populate the catalog with the package definition for your extension by running the `opm init` command:
+
[source,terminal]
----
$ opm init <extension_name> \
    --output json \
    > <catalog_dir>/index.json
----
+

> This command generates an `olm.package` declarative config blob in the specified catalog configuration file.

Reviewer: (Unless this is normal terminology for this context)
```suggestion
This command generates an `olm.package` declarative configuration blob in the specified catalog configuration file.
```

Author: Normally I would per IBM SG, but leaving "config blob"s alone for now to match `opm` CLI help output elsewhere. Will need to address in the CLI before potentially switching/syncing up to "configuration blob".

[id="olmv1-adding-plain-bundle-to-fbc_{context}"]
= Adding a plain bundle to a file-based catalog

> Currently, the `opm render` command does not support adding plain bundles to catalogs. You must manually add plain bundles to your file-based catalog, as shown in the following procedure.

Reviewer: ```suggestion
The `opm render` command does not support adding plain bundles to catalogs. You must manually add plain bundles to your file-based catalog, as shown in the following procedure.
```

.Procedure

> . Verify that your catalog's `index.json` or `index.yaml` file is similar to the following example:

Reviewer: Coul avoid possessive by flipping it
```suggestion
. Verify that `index.json` or `index.yaml` file for your catalog is similar to the following example:
```

.Verification

* Open your `index.json` or `index.yaml` file and ensure it is similar to the following example:
+
.Example `<catalog_dir>/index.json` file
[source,json]
----
{
    "schema": "olm.package",
    "name": "example-extension",
}
{
    "schema": "olm.bundle",
    "name": "example-extension.v0.0.1",
    "package": "example-extension",
    "image": "quay.io/rashmigottipati/example-extension-bundle:v0.0.1",
    "properties": [
        {
            "type": "olm.package",
            "value": {
            "packageName": "example-extension",
            "version": "v0.0.1"
            }
        },
        {
            "type": "olm.bundle.mediatype",
            "value": "plain+v0"
        }
    ]
}
{
    "schema": "olm.channel",
    "name": "preview",
    "package": "example-extension",
    "entries": [
        {
            "name": "example-extension.v0.0.1"
        }
    ]
}
----

[id="olmv1-publishing-fbc_{context}"]
= Building and publishing a file-based catalog

.Procedure

> . Build your file-bsaed catalog as an image by running the following command:

Reviewer: ```suggestion
. Build your file-based catalog as an image by running the following command:
```

// Module included in the following assemblies:
//
// * operators/olm_v1/olmv1-managing-catalogs.adoc

:_content-type: CONCEPT

[id="olmv1-building-plain-bundle-image-source_{context}"]
= Building a plain bundle image from an image source

> The Operator Controller currently only supports installing plain bundles created from a _plain bundle image_.

Reviewer 4: In the upstream doc, I did not follow ISG guidance on the usage of "only". 

Just to clarify, it should read something like: Currently, the Operator Controller supports installing plain bundles created only from a _plain bundle image_. 

Or maybe reworked to avoid only, but to get the same idea across. There are plans to support plain bundles for different images and I think Git repos on the road map.
